### PR TITLE
Forward port package puller API refactor from 7.x

### PR DESCRIPTION
**Pull request by a-palchikov**
## Description

Forward port changes to app/package puller APIs from 7.x as a continuation of forward-porting upgrade-via. This step also adds `AdditionalDependencies` to the relevant types to list additional package/app dependencies pulled when building an installer with support for upgrade-via.

## Type of change

* Internal change (not necessarily a bug fix or a new feature)

## Linked tickets and other PRs

* Refs https://github.com/gravitational/gravity/pull/1920

## TODOs

- [x] Self-review the change
- [x] Write tests
- [ ] Perform manual testing
- [ ] Address review feedback
- [ ] Update upstream references / tags / versions after upstream PR merges (linked above)

## Testing done

Added an integration test for the upload with dependency packages during upgrade.
Tested on 3-node cluster in following scenarios:
  * Installation
  * Upgrade from 9.0.0-beta.5.6 (commit 9b71d2b27084)

### CODE REVIEW

File: lib/app/pull.go [hunk 1]
```diff
@@ -0,0 +1,354 @@
+/*
+Copyright 2019 Gravitational, Inc.
```
Comment (wadells) [inline lib/app/pull.go:2]: 2021

File: lib/app/dependency_test.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2019 Gravitational, Inc.
```
Comment (wadells) [inline lib/app/dependency_test.go:2]: 2021

**Review by wadells — COMMENTED** [review]
Did you already take this through 8.0.x?  I'm confused about the flow of ports related to this change.

Are there any 9.0 specific details that need attention, or is this more or less verbatim from 7.0?

I invested ~30 minutes worth of review today, but @ ~1000 SLOC, that amounts largely to a scan of the changes.  I can put in further review if there are certain parts you're worried about.  But without direction, it'd mostly be learning the underlying architecture for and preexisting code for hours -- probably not high ROI.

**Comment by a-palchikov** [comment]
@wadells 

> Did you already take this through 8.0.x? I'm confused about the flow of ports related to this change.
> 
> Are there any 9.0 specific details that need attention, or is this more or less verbatim from 7.0?
> 
> I invested ~30 minutes worth of review today, but @ ~1000 SLOC, that amounts largely to a scan of the changes. I can put in further review if there are certain parts you're worried about. But without direction, it'd mostly be learning the underlying architecture for and preexisting code for hours -- probably not high ROI.

The changes are mostly verbatim copy from 7.x. I plan to not port these to 8.x if that's what you meant by taking it through 8.x. I've also tested the complete set of changes by upgrading 7.0.x to 9.x via 8.x using these changes if you meant that instead. I'm mostly worried the whole changeset is too large for a single PR hence I've been trying to split it in digestible chunks.
The changeset also includes changes from other PRs that make sense to port such as [custom planet images](https://github.com/gravitational/gravity/pull/1982) which was also only partially carried over.
I will try and draw your attention to deviations if there are any specifically, but I've been trying to cover the submissions with unit tests instead so this is largely unnecessary.

And sorry for the amount of changes - the ports are hopelessly overdue.

**Pull request closed without merging**